File: ts/algorithm/sorting/merge.ts
export function Merge (nums: Array<number>): Array<number> {
  return sort(nums)
}

function sort (nums: Array<number>): Array<number> {
  if (nums.length <= 1)
    return nums

  let middle = Math.floor(nums.length / 2)
  let idx = 0
  let left: Array<number> = []
  for (; idx < middle; idx++)
    left.push(nums[idx])
  let right: Array<number> = []
  for (; idx < nums.length; idx++)
    right.push(nums[idx])

  return merge(sort(left), sort(right))
}

function merge (left: Array<number>, right: Array<number>): Array<number> {
  let result: Array<number> = []

  while ((left.length > 0) && (right.length > 0)) {
    if (left[0] < right[0]) {
      result.push(left[0])
      left = left.slice(1)
    } else {
      result.push(right[0])
      right = right.slice(1)
    }
  }

  result = result.concat(left)
  result = result.concat(right)

  return result
}

File: python/algorithm/sorting/merge.py
import math

def sort(nums):
  if len(nums) <= 1:
    return nums

  middle = len(nums) // 2
  left = nums[:middle]
  right = nums[middle:]

  return merge(sort(left), sort(right))

def merge(left, right):
  l = 0
  r = 0
  result = []

  while l < len(left) and r < len(right):
    if left[l] < right[r]:
      result.append(left[l])
      l += 1
    else:
      result.append(right[r])
      r += 1

  while l < len(left):
    result.append(left[l])
    l += 1
  while r < len(right):
    result.append(right[r])
    r += 1

  return result
